src/main/java/ru/yandex/practicum/filmorate/service/FilmService.java

```diff
@@ -37,56 +71,69 @@ public Film findById(int id) {
+        // Важно: писать лайк в БД, а не только в память
+        filmStorage.addLike(filmId, userId);
     }
 
-    // Удалить лайк у фильма (теперь через Film)
     public void removeLike(int filmId, int userId) {
-        Film film = findById(filmId); // 404 если нет фильма
-        if (userStorage.findById(userId) == null) {
-            throw new NotFoundException("Пользователь не найден");
-        }
-        film.removeLike(userId);
+        // Проверяем наличие фильма/пользователя, чтобы вернуть 404 при отсутствии
+        findById(filmId);
+        userStorage.findById(userId);
+        // Важно: удалять лайк из БД
+        filmStorage.removeLike(filmId, userId);
     }
 
-    // Получить популярные фильмы
     public List<Film> getPopularFilms(int count) {
-        return filmStorage.findAll().stream()
-                .sorted(Comparator.comparingInt((Film f) -> f.getLikes().size()).reversed())
-                .limit(count)
-                .collect(Collectors.toList());
+        if (count <= 0) count = 10; // простая защита
+        // Критично: использовать выборку из БД с ORDER BY COUNT(likes) DESC
+        return filmStorage.getPopularFilms(count);
     }
 
-    // Обновление фильма с пользовательской валидацией
     public Film updateFilmCustomValidation(Film updatedFilm) {
         if (updatedFilm.getId() <= 0) {
             throw new ValidationException("ID фильма обязателен для обновления.");
         }
 
+        // Валидация существования фильма
         Film existingFilm = findById(updatedFilm.getId());
 
+        // Валидация и обновление MPA, если изменён
+        if (updatedFilm.getMpa() != null && updatedFilm.getMpa().getId() != 0) {
+            mpaStorage.findById(updatedFilm.getMpa().getId()); // NotFoundException если не найден
+            existingFilm.setMpa(updatedFilm.getMpa());
+        }
+
+        // Валидация и обновление жанров, если изменены
+        if (updatedFilm.getGenres() != null) {
+            for (Genre genre : updatedFilm.getGenres()) {
+                genreStorage.findById(genre.getId());
+            }
+            existingFilm.setGenres(updatedFilm.getGenres());
+        }
+
         if (updatedFilm.getName() != null && !updatedFilm.getName().isBlank()) {
             existingFilm.setName(updatedFilm.getName());
         }
-
         if (updatedFilm.getDescription() != null && updatedFilm.getDescription().length() <= 200) {
             existingFilm.setDescription(updatedFilm.getDescription());
         }
-
         if (updatedFilm.getReleaseDate() != null && !updatedFilm.getReleaseDate().isBefore(CINEMA_BIRTHDAY)) {
             existingFilm.setReleaseDate(updatedFilm.getReleaseDate());
         }
-
         if (updatedFilm.getDuration() > 0) {
             existingFilm.setDuration(updatedFilm.getDuration());
         }
 
         return filmStorage.updateFilm(existingFilm);
     }
+
+
+    public Film getFilmById(int id) {
+        return findById(id);
+    }
```

Comment:
Стоит сделать метод приватным, раз он используется только внутри других методов этого класса. Не очень понятен смысл этого метода, если он просто вызывает `findById()`

(It's worth making the method private since it's only used inside other methods of this class. It's not very clear what the point of this method is if it just calls `findById()`)